pkg/openstack/loadbalancer_test.go

```diff
@@ -593,3 +595,296 @@ func Test_getIntFromServiceAnnotation(t *testing.T) {
+			assert.Equal(t, got, tt.want)
+		})
+	}
+}
+
+func Test_getSubnetID(t *testing.T) {
+	type args struct {
+		svcConf *serviceConfig
+		service *corev1.Service
+		lbaasV2 *LbaasV2
+	}
+	tests := []struct {
+		name        string
+		args        args
+		want        string
+		expectedErr error
+	}{
+		{
+			name: "test get subnet from service annotation",
+			args: args{
+				svcConf: &serviceConfig{},
+				lbaasV2: &LbaasV2{
+					LoadBalancer{
+						opts: LoadBalancerOpts{
+							LBClasses: map[string]*LBClass{
+								"test-class": {
+									SubnetID:           "test-id",
+									FloatingNetworkID:  "test-floating-network-id",
+									FloatingSubnetID:   "test-floating-subnet-id",
+									FloatingSubnet:     "test-floating-subnet",
+									NetworkID:          "test-network-id",
+									FloatingSubnetTags: "test-floating-subnet-tags",
+									MemberSubnetID:     "test-member-subnet-id",
+								},
+							},
+						},
+					},
+				},
+				service: &corev1.Service{
+					ObjectMeta: v1.ObjectMeta{
+						Annotations: map[string]string{
+							"loadbalancer.openstack.org/subnet-id": "test-id",
+							"loadbalancer.openstack.org/class":     "test-class",
+						},
+					},
+				},
+			},
+			want: "test-id",
+		},
+		{
+			name: "test get subnet from config class",
+			args: args{
+				svcConf: &serviceConfig{},
+				lbaasV2: &LbaasV2{
+					LoadBalancer{
+						opts: LoadBalancerOpts{
+							LBClasses: map[string]*LBClass{
+								"test-class": {
+									SubnetID:           "test-class-subnet-id",
+									FloatingNetworkID:  "test-floating-network-id",
+									FloatingSubnetID:   "test-floating-subnet-id",
+									FloatingSubnet:     "test-floating-subnet",
+									NetworkID:          "test-network-id",
+									FloatingSubnetTags: "test-floating-subnet-tags",
+									MemberSubnetID:     "test-member-subnet-id",
+								},
+							},
+						},
+					},
+				},
+				service: &corev1.Service{
+					ObjectMeta: v1.ObjectMeta{
+						Annotations: map[string]string{
+							"loadbalancer.openstack.org/class": "test-class",
+						},
+					},
+				},
+			},
+			want:        "test-class-subnet-id",
+			expectedErr: nil,
+		},
+		{
+			name: "test get subnet from config class with invalid loadbalancer class",
+			args: args{
+				svcConf: &serviceConfig{},
+				lbaasV2: &LbaasV2{
+					LoadBalancer{
+						opts: LoadBalancerOpts{
+							SubnetID: "test-subnet-id",
+						},
+					},
+				},
+				service: &corev1.Service{
+					ObjectMeta: v1.ObjectMeta{
+						Annotations: map[string]string{
+							"loadbalancer.openstack.org/class": "test-class",
+						},
+					},
+				},
+			},
+			want:        "",
+			expectedErr: fmt.Errorf("invalid loadbalancer class %q", "test-class"),
+		},
+		{
+			name: "test get subnet from default config",
+			args: args{
+				svcConf: &serviceConfig{},
+				lbaasV2: &LbaasV2{
+					LoadBalancer{
+						opts: LoadBalancerOpts{
+							SubnetID: "test-default-subnet-id",
```

Comment:
I'd add a LBClass here as well, just to make sure it's ignored in case the class is not set below. What do you think?

Reply:
sounds nice but looking at the function the default config is only selected when the config class is empty, so if I add the extra class, it's supposed to pick it up and not ignore it

Reply:
if you look at the screenshot, if the configClassName is not set to ""(empty string) the default config class is not picked, if I add a LbClass here the default config class won't be ignored, and we're testing for the default config class, if you'd like I can add an extra test case with LbClass set and the default  config class set, it should ignore the default config class and pick the LBClass
<img width="725" alt="Screenshot 2023-10-10 at 13 27 20" src="https://github.com/kubernetes/cloud-provider-openstack/assets/69452689/eee838d6-0a20-434e-88b2-91413ccc1285">

Reply:
i misunderstood you before, I've added it in now